### PR TITLE
Adds FirehoseHandler for customization, pruning and aggregation

### LINKED ISSUE #557: Support "firehose mode"
Full description pending, but this will describe a mode where brave collects data regardless of sampling decision. At the point where spanReporter is used today, a branch will occur where "firehose" reporter gets all spans, and the existing one, only sampled ones. This is in support of aggregators such as span metrics.

Coming out of https://github.com/openzipkin/brave/issues/529

### PR DESCRIPTION
**Pull request by adriancole**
This is adds the ability to add `FirehoseHandler` instances to customize data before it is reported to Zipkin. It accepts a pair of (TraceContext, MutableSpan) which allows you to inspect other data such as headers.

The important piece is that this can receive data when a context isn't sampled for Zipkin, or if the zipkin reporter is NoOp. This applies when `TraceContext.sampledLocal()` is set explicitly or indirectly via `Firehose.alwaysSampleLocal()`

This can be used for metrics handlers such as micrometer or aggregation feeds like haystack. Examples to follow.

Fixes #557 

cc @drolando who has done similar efforts in python (py_zipkin)

### CODE REVIEW

File: brave/src/main/java/brave/RealSpan.java [hunk 1]
```diff
@@ -12,22 +13,22 @@
...
+  final Firehose firehose;
   final ErrorParser errorParser;
   final RealSpanCustomizer customizer;
 
   RealSpan(TraceContext context,
       PendingSpans pendingSpans,
       MutableSpan state,
       Clock clock,
-      SpanReporter spanReporter,
+      SpanReporter firehose,
```
Comment (devinsba) [inline brave/src/main/java/brave/RealSpan.java]: I see SpanReporter is a subtype of Firehose but did you intend to not change the type here?

File: brave/src/main/java/brave/internal/InternalPropagation.java [hunk 2]
```diff
@@ -11,11 +11,26 @@
  * <p>Inspired by {@code okhttp3.internal.Internal}.
  */
 public abstract class InternalPropagation {
+  public static final int FLAG_SAMPLED = 1 << 1;
+  public static final int FLAG_SAMPLED_SET = 1 << 2;
+  public static final int FLAG_DEBUG = 1 << 3;
+  public static final int FLAG_SHARED = 1 << 4;
+  public static final int FLAG_SAMPLED_LOCAL = 1 << 5;
```
Comment (devinsba) [inline brave/src/main/java/brave/internal/InternalPropagation.java]: Might be good to get a contextual comment about these flags in here

**Comment by adriancole** [comment]
thanks for the brave first look!

will work in your comments on next commit. I also will change a little
internals to try and reduce overhead a bit. most importantly I will add a
learning test

On 20 Sep 2018 03:29, "Brian Devins-Suresh" <notifications@github.com>
wrote:

*@devinsba* commented on this pull request.
------------------------------

In brave/src/main/java/brave/RealSpan.java
<https://github.com/openzipkin/brave/pull/791#discussion_r219110090>:

>    final ErrorParser errorParser;
   final RealSpanCustomizer customizer;

   RealSpan(TraceContext context,
       PendingSpans pendingSpans,
       MutableSpan state,
       Clock clock,
-      SpanReporter spanReporter,
+      SpanReporter firehose,

I see SpanReporter is a subtype of Firehose but did you intend to not
change the type here?
------------------------------

In brave/src/main/java/brave/internal/InternalPropagation.java
<https://github.com/openzipkin/brave/pull/791#discussion_r219112079>:

> @@ -11,11 +11,26 @@
  * <p>Inspired by {@code okhttp3.internal.Internal}.
  */
 public abstract class InternalPropagation {
+  public static final int FLAG_SAMPLED = 1 << 1;
+  public static final int FLAG_SAMPLED_SET = 1 << 2;
+  public static final int FLAG_DEBUG = 1 << 3;
+  public static final int FLAG_SHARED = 1 << 4;
+  public static final int FLAG_SAMPLED_LOCAL = 1 << 5;

Might be good to get a contextual comment about these flags in here

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/openzipkin/brave/pull/791#pullrequestreview-157202517>,
or mute the thread

File: brave/src/main/java/brave/internal/InternalPropagation.java (hunk 2 above)
Comment (adriancole) [reply brave/src/main/java/brave/internal/InternalPropagation.java]: done

File: brave/src/main/java/brave/RealSpan.java (hunk 1 above)
Comment (adriancole) [reply brave/src/main/java/brave/RealSpan.java]: added a commit to change names etc. Now the router thing is called FirehoseDispatcher and it doesn't implement Firehose.. makes it easier to see what's doing what.

**Comment by adriancole** [comment]
Added FirehoseDispatcher to clarify the difference between dispatch and implementation

File: brave/src/main/java/brave/internal/recorder/Firehose.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
+package brave.internal.recorder;
+
+import brave.Span;
+import brave.Tracing;
+import brave.propagation.TraceContext;
+
+public interface Firehose {
```
Comment (adriancole) [inline brave/src/main/java/brave/internal/recorder/Firehose.java]: This is the main type

**Comment by adriancole** [comment]
ps Firehose or FirehoseHandler? cc @devinsba

**Comment by adriancole** [comment]
ok the api is in shape now, though it needs some polish before merge. I plan to add a example commit on top of this, and welcome others to give it a try though no pressure.

**Pull request merged**